File: research/phase3/task85c-i/reference/verify_jobid_e2.mjs
#!/usr/bin/env node
import fs from "node:fs";
import path from "node:path";
import crypto from "node:crypto";
const out=path.resolve(path.dirname(new URL(import.meta.url).pathname),"..");
const f=JSON.parse(fs.readFileSync(path.join(out,"fixtures/G1V2_E2_JOBID_FIXTURE.json"),"utf8"));
const canonical=x=>Buffer.from(JSON.stringify(Object.fromEntries(Object.keys(x).sort().map(k=>[k,x[k]])))+"\n","utf8");
const jobid=x=>"j-"+crypto.createHash("sha256").update(Buffer.concat([Buffer.from("G1V2-JOB\0"),canonical(x)])).digest("hex").slice(0,40);
for(const version of ["v1_1","v1_2"]){
  if(canonical(f[version].payload).toString("hex")!==f[version].canonical_payload_hex) throw new Error(`canonical ${version}`);
  if(jobid(f[version].payload)!==f[version].jobid) throw new Error(`jobid ${version}`);
}
if(f.v1_1.jobid===f.v1_2.jobid) throw new Error("cross-version collision");
console.log("JOBID_E2_JAVASCRIPT=PASS");
console.log("CROSS_IMPLEMENTATION_BYTE_IDENTITY=PASS");

File: research/phase3/task85c-i/reference/verify_jobid_e2.py
#!/usr/bin/env python3
"""Independent Python implementation of the E2 JobID fixture."""
import hashlib
import json
import unicodedata
from pathlib import Path

OUT=Path(__file__).resolve().parent.parent
def norm(x):
    if isinstance(x,str): return unicodedata.normalize("NFC",x)
    if isinstance(x,list): return [norm(v) for v in x]
    if isinstance(x,dict): return {k:norm(x[k]) for k in sorted(x)}
    return x
def canon(x): return (json.dumps(norm(x),ensure_ascii=False,sort_keys=True,separators=(",",":"))+"\n").encode()
def jobid(x): return "j-"+hashlib.sha256(b"G1V2-JOB\0"+canon(x)).hexdigest()[:40]
f=json.loads((OUT/"fixtures/G1V2_E2_JOBID_FIXTURE.json").read_text())
for version in ("v1_1","v1_2"):
    assert canon(f[version]["payload"]).hex()==f[version]["canonical_payload_hex"]
    assert jobid(f[version]["payload"])==f[version]["jobid"]
assert f["v1_1"]["jobid"]!=f["v1_2"]["jobid"]
e2=json.loads((OUT/"G1V2_EXECUTION_IDENTITY_ERRATUM_E2.json").read_text())
assert e2["jobid"]["dependency_field"]=="dependency_job_ids"
assert e2["scientific_boundary"]["scientific_identity_depends_on_blind_id"] is False
assert e2["scientific_boundary"]["scientific_identity_depends_on_escrow_key"] is False
print("JOBID_E2_PYTHON=PASS")
print("H_SC02=CLOSED")
